src/sds.c

```diff
@@ -125,7 +130,7 @@ sds _sdsnewlen(const void *init, size_t initlen, int trymalloc) {
 sds sdswrite(char *buf, size_t bufsize, char type, const char *init, size_t initlen) {
     assert(bufsize >= sdsReqSize(initlen, type));
     int hdrlen = sdsHdrSize(type);
-    size_t usable = bufsize - hdrlen - 1;
+    size_t usable = min(bufsize - hdrlen - 1, sdsTypeMaxSize(type));
```

Comment:
We store usable in the `alloc` field in the sds header. If we round it down here, we may store something that's not the allocation size. It may affect things like memory usage calculation and what we pass to `zfree_with_size()`.

I struggled with this in the past an managed to avoid it somehow by not calling this function with sizes that hit this corner case.

Can we continue to avoid it and add an assert instead?

```suggestion
    size_t usable = bufsize - hdrlen - 1;
    assert(usable <= sdsTypeMaxSize(type));
```